File: 410-split-array-largest-sum/split-array-largest-sum.java
class Solution {
    public int splitArray(int[] nums, int k) {
        if(k == 1){
            return findSum(nums);
        }

        if(k == nums.length){
            return maxEle(nums);
        }

        int min_ans = maxEle(nums);
        int max_ans = findSum(nums);

        return binarySearch(nums, k, min_ans, max_ans);
    }

    public int maxEle(int[] arr){
        int max = Integer.MIN_VALUE;
        for(int ele:arr){
            if(max < ele){
                max = ele;
            }
        }

        return max;
    }

    public int findSum(int[] arr){
        int sum = 0;

        for(int ele:arr){
            sum += ele;
        }

        return sum;
    }

    public int binarySearch(int[] arr, int k, int start, int end){

        while(start < end){
            int mid = start + (end - start) / 2;
            int pieces = 1;

            int sum = 0;
            for(int ele:arr){
                sum += ele;
                if(sum > mid){
                    sum = ele;
                    pieces++;
                }
            }

            if(pieces <= k){
                end = mid;
            }else{
                start = mid + 1;
            }
        }

        return end;
    }

}

File: 410-split-array-largest-sum/split-array-largest-sum.cpp
class Solution {
public:
    bool check(vector<int>& nums, int k, int mid){
        int numsTotal = 0;
        int part = 1;

        for(int i = 0; i < nums.size(); i++){
            if(numsTotal + nums[i] <= mid){
                numsTotal += nums[i];
            }else{
                part++;
                if(part > k) return false;
                numsTotal = nums[i];
            }
        }

        return true;
    }

    int splitArray(vector<int>& nums, int k) {
        int low = *max_element(nums.begin(), nums.end());
        int high = 0;

        for(int i:nums){
            high += i;
        }

        while(low <= high){
            int mid = low + (high - low) / 2;

            if(check(nums, k, mid)){
                high = mid - 1;
            }else{
                low = mid + 1;
            }
        }

        return low;
    }
};